src/java_bytecode/java_types.cpp

```diff
@@ -167,6 +167,243 @@ exprt java_bytecode_promotion(const exprt &expr)
+/// \param class_name_prefix: The name of the class to use to prefix any found
+///   generic types
+void add_generic_type_information(
+  java_generic_typet &generic_type,
+  const std::string &parameters,
+  const std::string &class_name_prefix)
+{
+  PRECONDITION(parameters[0] == '<');
+  PRECONDITION(parameters[parameters.size() - 1] == '>');
+
+  // parse contained types, can be either type variables, starting with T
+  // or instantiated types
+  const std::vector<typet> &params =
+    parse_list_types(parameters, class_name_prefix, '<', '>');
+
+  CHECK_RETURN(params.size() > 0); // We should have at least one generic param
+
+  // take these types - they should either be java_generic_parameters, in which
+  // case they can be directly added to the generic_type
+  // otherwise they should be wrapped in a java_generic_inst_parametert
+
+  std::transform(
+    params.begin(),
+    params.end(),
+    std::back_inserter(generic_type.generic_type_variables()),
+    [](const typet &type) -> java_generic_parametert {
+      if(is_java_generic_parameter(type))
+      {
+        const java_generic_parametert &gen_type =
+          to_java_generic_parameter(type);
+        return gen_type;
+      }
+      else
+      {
+        INVARIANT(
+          is_reference(type), "All generic parameters should be references");
+        java_generic_inst_parametert inst_type(to_symbol_type(type.subtype()));
+        return inst_type;
+      }
+    });
+}
+
+/// Take a signature string and remove everything in angle brackets allowing
+/// for the type to be parsed normally.
+/// \param src: The input string
+/// \return The input string with everything between angle brackets removed
+std::string erase_type_arguments(const std::string &src)
+{
+  std::string class_name = src;
+  std::size_t f_pos = class_name.find('<', 1);
+
+  while(f_pos != std::string::npos)
+  {
+    std::size_t e_pos = find_closing_delimiter(class_name, f_pos, '<', '>');
+    if(e_pos == std::string::npos)
+      throw unsupported_java_class_signature_exceptiont("recursive generic");
+
+    // erase generic information between brackets
+    class_name.erase(f_pos, e_pos - f_pos + 1);
+
+    // Search the remainder of the string for generic signature
+    f_pos = class_name.find('<', e_pos + 1);
+  }
+  return class_name;
+}
+
+/// Returns the full class name, skipping over the generics
+/// \param src: a type descriptor or signature like LOuterClass<TT;>.Inner;
+/// \return The full name of the class like OuterClass.Inner
+std::string gather_full_class_name(const std::string &src)
+{
+  PRECONDITION(src[0] == 'L');
+  PRECONDITION(src[src.size() - 1] == ';');
+
+  std::string class_name = src.substr(1, src.size() - 2);
+
+  class_name = erase_type_arguments(class_name);
+
+  std::replace(class_name.begin(), class_name.end(), '.', '$');
+  std::replace(class_name.begin(), class_name.end(), '/', '.');
+  return class_name;
+}
+
+size_t find_closing_semi_colon_for_reference_type(
```

Comment:
why not in the header?

Reply:
As above - let me know if you'd be happy with my proposed solution

Reply:
ok with me